lib/puppet/type/tomcat_install.rb

```diff
@@ -0,0 +1,50 @@
+Puppet::Type.newtype(:tomcat_install) do
+  desc "Installs the software into the given directory."
+
+  ensurable do
+    defaultvalues
+    defaultto :present
+  end
+
+  autorequire(:user) do
+    self[:user]
+  end
+
+  autorequire(:group) do
+    self[:group]
+  end
+
+  newparam(:catalina_home, namevar: true) do
+    desc "The directory of the Tomcat installation."
+
+    validate do |value|
+      raise ArgumentError, "Catalina home is required." if value.nil?
+    end
+  end
+
+  newparam(:version) do
+    desc "Specifies the owner of the source installation directory."
+
+    validate do |value|
+      raise ArgumentError, "Tomcat version is required." if value !~ /\d{1,2}\.\d{1,2}\.\d{1,2}/
+    end
+  end
+
+  newparam(:user) do
+    desc "Specifies the owner of the source installation directory."
+
+    defaultto 'tomcat'
+  end
+
+  newparam(:group) do
+    desc "Specifies the group of the source installation directory."
+
+    defaultto 'tomcat'
+  end
+
+  validate do
+    unless self[:version]
```

Comment:
Always requiring a version would be backwards-incompatible with the previous usage of just passing a `source_url` and having the version be part of that. Could you add a `source_url` parameter that is used if no version is passed? That way internal file hosting can be used as well.

Reply:
Done.

Reply:
\o/